compiler/rustc_trait_selection/src/error_reporting/traits/fulfillment_errors.rs

```diff
@@ -1835,23 +1835,32 @@ impl<'a, 'tcx> TypeErrCtxt<'a, 'tcx> {
                     if impl_trait_ref.references_error() {
                         return false;
                     }
+                    let self_ty = impl_trait_ref.self_ty().to_string();
                     err.highlighted_help(vec![
                         StringPart::normal(format!(
                             "the trait `{}` ",
                             impl_trait_ref.print_trait_sugared()
                         )),
                         StringPart::highlighted("is"),
                         StringPart::normal(" implemented for `"),
-                        StringPart::highlighted(impl_trait_ref.self_ty().to_string()),
+                        if let [TypeError::Sorts(_)] = &terrs[..] {
+                            StringPart::normal(self_ty)
+                        } else {
+                            StringPart::highlighted(self_ty)
+                        },
                         StringPart::normal("`"),
                     ]);
 
                     if let [TypeError::Sorts(exp_found)] = &terrs[..] {
                         let exp_found = self.resolve_vars_if_possible(*exp_found);
-                        err.help(format!(
-                            "for that trait implementation, expected `{}`, found `{}`",
-                            exp_found.expected, exp_found.found
-                        ));
+                        err.highlighted_help(vec![
+                            StringPart::normal("for that trait implementation, "),
+                            StringPart::normal("expected `"),
+                            StringPart::highlighted(exp_found.expected.to_string()),
+                            StringPart::normal("`, found `"),
+                            StringPart::highlighted(exp_found.found.to_string()),
+                            StringPart::normal("`"),
+                        ]);
```

Comment:
Discussion: personally, I find that splitting `expected` and `actual` into two lines but then line them up to be faster to visually parse than having them inline as ``expected `()` but found `i32` ``. Of course, this will increase the line count.

```
    expected: `()`
    found:    `i32`
```

What do you think?

(I had this review comment on the wrong PR, https://github.com/rust-lang/rust/pull/132157#discussion_r1817691283)

Reply:
One the one hand, I agree (and we have a method for that formatting already that we can easily reuse), but I feel that these diagnostics are 1) already quite long and 2) with all of the changes taken in aggregate I feel like this line could be safely removed. The reason I added it in the first place was because we didn't have a good way to show the comparison of types that was expected and found. The way I look at it, if you take this one line in isolation you don't get enough information to fix the issue, but if you look at the previous lines ("T is not implemented for A, but it is implemented for B") you do, and with the highlighting added to that this line's function is dramatically reduced. As a compromise (with myself) I left the line without turning it multiline (as proposed), but I feel we could do without it entirely. What do you think?

Reply:
That seems fine to me as well :+1: